src/main/java/com/allenai/ml/linalg/SparseVector.java

```diff
@@ -87,5 +89,5 @@ public double dotProduct(Vector other) {
           result += val * other.at(idx);
```

Comment:
It can be that slow if you do a dot product between two sparse vectors, but the dot product case you deal with most often in ML is a dense weight vector against a sparse feature vector, which is linear, since you iterate over the `SparseVector` and do lookups agains the `double[]`-backed `DenseVector` in O(1). In this case the dimensionality is the same, but the feature vector has stored far fewer entries than the full dimension stored by the dense weight vector. 

Also, GSCollection has both a hash-based primitive map and sorted ones, so you can get either behavior depending on the map you pass in. Could also provide friendly factory functions to make this easier to see and use.